### ISSUE TITLE
Intersection type in template literal is not reduced to its bare type

### ISSUE THREAD

**Issue by unional** [microsoft/TypeScript#57918]
### 🔎 Search Terms

template literal, intersection type

### 🕗 Version & Regression Information

- This is a crash
- This changed between versions 5.0 and 5.1
- This changed in commit or PR #48034 #52345 #54648

### ⏯ Playground Link

https://www.typescriptlang.org/play?#code/C4TwDgpgBAysBOBLAdgcwCrggHgNJQgA9gJkATAZygoRVQD4oBeKfIk8qmpNKAMigoAZhHhQAqgCgAkAH4oAbVwAaCQF0CxUpUXjVuNTPkBvAL6aOO44oAKg5FAAGAEmO5TjtQC4oAV2QA1sgA9gDuDqZGUADkJAC2YAA2AIYkADKIJPDJidEyPtHcdBlZOXnSBUVo5T7IEABuopKSoJBQAEoAjMywtGiYkNhVDJIA9KNQkwB6si1YHQBMPXA8GFhDfaj8UNbJPt2m9GMT07Nzbe0AzMubAzjRyQBGAMbRR+OTUDPn0O0ALDdVndsA8XtFtrt9lBDsdPt8fh0AKyAujAlzGZC+OKPUQed4nL6zVq-ABsKP663RmOxuMcEKgeygB3xcLOUCAA

### 💻 Code

```ts
type StringType<K extends string> = K extends string & infer U
	? [K, U] extends [U, K]
	? {} extends { [P in `${K}`]: unknown }
	? 'templateLiteral'
	: 'stringLiteral'
	: 'string'
	: never

type R1 = StringType<string>
//   ^? type R1 = "string"
type R2 = StringType<string & { a: 1 }>
//   ^? type R2 = "string"

type R3 = StringType<'abc'>
//   ^? type R3 = "stringLiteral"
type R4 = StringType<'abc' & { a: 1 }>
//   ^? type R4 = "templateLiteral" <-- should be "stringLiteral"

type R5 = StringType<`${number}`>
//   ^? type R5 = "templateLiteral"
type R6 = StringType<`${number}` & { a: 1 }>
//   ^? type R6 = "templateLiteral"
```

### 🙁 Actual behavior

```ts
type R = `${'abc' & { a: 1 }}`
```

is not reduced to `${'abc'}` => `abc`

### 🙂 Expected behavior

```ts
type R = `${'abc' & { a: 1 }}`
```

should be reduced to `${'abc'}` => `abc`

### Additional information about the issue

I mentioned this in #54648 after it is closed. It is limiting our ability to write the types that works with string literal and template literal and there is no alternative way to workaround that.
I'm suggesting this issue should be fixed and restore the behavior in 5.0.

Here is my original comment:

This behavior is causing a few types in `type-plus` to fail (e.g. `IsTemplateLiteral`, `IsStringLiteral`, `Omit`, `IsNegative`, etc) https://github.com/unional/type-plus/issues/429.

In term of soundness, IMO it does make sense that `${string & { a: 1 }}` to be reduced to `${string}`.

in JS, it would be:

```js
const extendedStr = Object.assign('abc', { a: 1 })
console.log(`${extendedStr}`) // 'abc'
```

the reasoning being the `toString(): string` remains unchanged thus the resulting type should be safe to reduce.

**Comment by jcalz** [comment]
> * This is a crash

No, the tsc process didn't crash.

**Comment by unional** [comment]
> No, the tsc process didn't crash.

Correct. Updated. 🍻

**Comment by HansBrende** [comment]
@RyanCavanaugh this is the first type I've seen that is **not identical to itself** via `isTypeIdenticalTo`:

```ts
export type Equals<X, Y> =
    (<T>() => T extends X ? 1 : 2) extends
    (<T>() => T extends Y ? 1 : 2) ? true : false;

type EqualsItself = 
//   ^?  type EqualsItself = false
    Equals<
        `${'abc' & { a: 1 }}`,
        `${'abc' & { a: 1 }}`
    >
```

[Playground link](https://www.typescriptlang.org/play/?#code/KYDwDg9gTgLgBDAnmYcCiBHArgQwDYDOAPABoA0cAmgHxwC8AUHM3ABREAq1rAlPbRzigYwAHYATAnBJwA-HACMcAFxwATH2FjJTFuy69+cQVolTKcxSvV95MKFlSqAZvgLAA3AwZIU6bG4AkjDueM70cAwA9FEsAHqyusyYuIRESSzMAAYAJADeAOQ4AEYAxgVwAGRweXA4qkoAvo1ZZBmZuYUl5VU1dQ1wzVkZ1AxAA)

That wasn't always the case: in v5.0.4 we have the correct `type EqualsItself = true`.

So maybe there is something unintentional going on here? I'm unsure if it's covered by this existing issue or it's an entirely different issue, but it does seem like there is something broken, as many type-testing libraries rely on this `Equals` implementation to determine if two arbitrary types are identical (and a copy-pasted type definition should surely be identical in every possible sense).

**Comment by RyanCavanaugh** [comment]
The `Equals` type there is (*intentionally**) reliant on internal compiler implementation details and shouldn't be used unless you're actively trying to break something. Its use in type testing libraries is bad and I would discourage its use.

**Comment by HansBrende** [comment]
@RyanCavanaugh I agree with your general sentiment, however, that's not quite what I'm trying to get across. Let me simplify and refocus my train of thought by taking out the `Equals` implementation altogether, as it is somewhat of a red herring:

```typescript
type ExtendsItself = 
//   ^?  type ExtendsItself = false
    (<T>() => T extends `${'a' & { a: 1 }}` ? 1 : 2) extends 
    (<T>() => T extends `${'a' & { a: 1 }}` ? 1 : 2) ? true : false;
```

If this should _not_ be considered a regression, then we have the seemingly strange situation above where **there exists a type that does not extend itself**, similar to some kind of `NaN`.

Which begs the question: do you consider it a regression, or not? I think the answer to that question will really help to clarify the situation.

**Comment by RyanCavanaugh** [comment]
Internal implementation details are subject to change at any time. `Equals` *may* evaluate to `true` but there are no guarantees on this -- it's dependent on whether or not we choose to intern a given type, which is a performance trade-off that may be situationally-dependent.

**Comment by HansBrende** [comment]
Or, as they say, a picture's worth a thousand words 😅:

<img width="929" alt="Image" src="https://github.com/user-attachments/assets/5c59cd28-d4e7-4603-9b71-8b4769e30e33" />

[Playground link](https://www.typescriptlang.org/play/?#code/FAUwHgDg9gTgLgAgDYkQQwFwIBQB4AqAfNgJQIC8hC+C4cIAdgCYDOCABgCQDeA5GrwQAyBNwSYEARgQBfGewQB+KQiwAmMuQQMArkiQBCANyhIsRCkQAjLHiKkKVGnUasOPfoJFiJ0uQuVpdU1tPUMTYDQKBCsjIA)

**Comment by HansBrende** [comment]
Ok, I went ahead and opened a [separate issue](https://github.com/microsoft/TypeScript/issues/61098) for the above, since I believe it could be resolved separately from this current issue.

**Comment by RyanCavanaugh** [comment]
> What I want to know is: would you consider that a regression or not?

No.

**Comment by HansBrende** [comment]
Update on my above comments: I fixed that slightly tangential issue in https://github.com/microsoft/TypeScript/pull/61113, which was merged a few days ago (incidentally, also fixing that particular problem with the `Equals` hack).

So to re-focus on the issue at hand: personally, I'm ambivalent whether tagged strings should have their tags removed in a template string. 

However, **the big issue I do see here** is that there is _no way_ to remove tags from a literal if you don't know the _exact_ type of the tag(s) in advance.

I.e., it is utterly impossible to write an `Untag<T>` type. For any function you come up with, I can give you a literal that can't be untagged. Here's a good test-case:

```typescript
type Evil = 1234 & {a: 0 | 1} & {a: 1 | 2} & {toString: Number['toString']}

type YourFailedAttempt = Untag<Evil>
```

This means that libraries that do conditional string manipulation, etc. will necessarily fail if the user passes in any tagged primitive. It would be really nice to have some way of stripping arbitrary tags from an arbitrary primitive literal.
